cmd/gpu_plugin/gpu_plugin_test.go

```diff
@@ -109,37 +126,47 @@ func TestScan(t *testing.T) {
 	}
 
 	for _, tc := range tcases {
-		t.Run(tc.name, func(t *testing.T) {
-			for _, devfsdir := range tc.devfsdirs {
-				if err := os.MkdirAll(path.Join(devfs, devfsdir), 0750); err != nil {
-					t.Fatalf("Failed to create fake device directory: %+v", err)
-				}
+		scanWithParams(t, tc, sysfs, devfs, 1)
+		scanWithParams(t, tc, sysfs, devfs, 4)
+	}
+}
+
+func scanWithParams(t *testing.T, tc testCase, sysfs, devfs string, shareCount int) {
+	name := fmt.Sprintf("%s-%d", tc.name, shareCount)
+	t.Run(name, func(t *testing.T) {
+		for _, devfsdir := range tc.devfsdirs {
+			if err := os.MkdirAll(path.Join(devfs, devfsdir), 0750); err != nil {
+				t.Fatalf("Failed to create fake device directory: %+v", err)
 			}
-			for _, sysfsdir := range tc.sysfsdirs {
-				if err := os.MkdirAll(path.Join(sysfs, sysfsdir), 0750); err != nil {
-					t.Fatalf("Failed to create fake device directory: %+v", err)
-				}
+		}
+		for _, sysfsdir := range tc.sysfsdirs {
+			if err := os.MkdirAll(path.Join(sysfs, sysfsdir), 0750); err != nil {
+				t.Fatalf("Failed to create fake device directory: %+v", err)
 			}
-			for filename, body := range tc.sysfsfiles {
-				if err := ioutil.WriteFile(path.Join(sysfs, filename), body, 0600); err != nil {
-					t.Fatalf("Failed to create fake vendor file: %+v", err)
-				}
+		}
+		for filename, body := range tc.sysfsfiles {
+			if err := ioutil.WriteFile(path.Join(sysfs, filename), body, 0600); err != nil {
+				t.Fatalf("Failed to create fake vendor file: %+v", err)
 			}
+		}
 
-			plugin := newDevicePlugin(sysfs, devfs, 1)
+		plugin := newDevicePlugin(sysfs, devfs, shareCount)
 
-			notifier := &mockNotifier{
-				scanDone: plugin.scanDone,
-			}
+		notifier := &mockNotifier{
+			scanDone: plugin.scanDone,
+		}
 
-			err := plugin.Scan(notifier)
-			// Scans in GPU plugin never fail
-			if err != nil {
-				t.Errorf("unexpected error: %+v", err)
-			}
-			if tc.expectedDevs != notifier.devCount {
-				t.Errorf("Wrong number of discovered devices")
-			}
-		})
-	}
+		err := plugin.Scan(notifier)
+		// Scans in GPU plugin never fail
+		if err != nil {
+			t.Errorf("unexpected error: %+v", err)
+		}
+		if shareCount*tc.expectedDevs != notifier.devCount {
+			t.Errorf("Expected %d, discovered %d devices",
+				tc.expectedDevs, notifier.devCount)
+		}
+		// remove dirs/files for next test
+		os.RemoveAll(sysfs)
```

Comment:
Alternatively `root`, `sysfs` and `devfs` could be defined and created inside the function with `root` removal deferred until the function's exit.

Reply:
Hm. That way separate sysfs & devs removal wouldn't be needed, one could just remove rootfs => sounds good.

Would it be OK as commit on top of the PR, or do you want it to be rebased to some specific commit in the PR?

Reply:
Usually we squash all commits into one before merging, so either way is ok.